### ISSUE TITLE
The compiler should not report CS8347 when there is no danger of escape

### ISSUE THREAD

**Issue by Dancpaz** [dotnet/roslyn#61296]
Apologies if I have misunderstood this error or why it is occurring.

**Expected:**
When there is no danger of a reference to a local method's variable escaping the method, the compiler does not report CS8347.

**Actual:**
The compiler reports CS8347 even when there is no danger of escape and more than one way for the compiler to guarantee that.

### Details
In the `GetThing<T>` method of the following example code, the compiler complains that the ref we are returning might reference a local variable we have created for our call to the `GetThingAndIndex<T>` method.  Generic type constraints render that impossible, and even without the type constraints, there is no logical path for such a reference to be returned.

This error can be eliminated by turning the `GetThing<T>` method into a clone of `GetThingAndIndex<T>` without the out parameter, which seems to emphasize that there is no danger here.

Once again, I apologize if I have misunderstood CS8347 or the mechanics by which such an escape might reasonably occur.

### Platform Details
.NET 6
Visual Studio 17.0.0 64-bit
Windows 11

### Example
```
public interface IThingController
{
    interface IThing { }

    bool GetThingIndex(int id, out int index);

    ref T GetThingByIndex<T>(int index)
        where T : class, IThing;

    ref T GetThingAndIndex<T>(int id, out int index)
        where T : class, IThing
    {
        if(!GetThingIndex(id, out index))              // here the out is sent to a black box, but no ref is generated
            throw new KeyNotFoundException();

        return ref GetThingByIndex<T>(index);   // here the out is copied, so the resulting ref can't be referencing it
    }

    ref T GetThing<T>(int id)
        where T : class, IThing
    {
        return ref GetThingAndIndex<T>(id, out var _);
        // Error	CS8347	Cannot use a result of 'IThingController.GetThingAndIndex<T>(int, out int)' in this context
        //  because it may expose variables referenced by parameter 'index' outside of their declaration scope
    }
}
```

**Comment by jcouv** [comment]
@cston Could you take a look and advise/triage? Thanks

**Comment by RikkiGibson** [comment]
I think the code producing the diagnostic here is equivalent to the following:

```cs
    ref T GetThing<T>(int id)
        where T : class, IThing
    {
        int unused;
        return ref GetThingAndIndex<T>(id, out unused);
        // Error	CS8347	Cannot use a result of 'IThingController.GetThingAndIndex<T>(int, out int)' in this context
        //  because it may expose variables referenced by parameter 'index' outside of their declaration scope
    }
```

I think this will be fixed in C# 11, since we will make `out` parameters `scoped` by default.
